### PR TITLE
Join order benchmark (IMDB) validation and benchmark improvements 

### PR DESCRIPTION
**Pull request by Tmonster**
This PR has 2 separate improvements, and basically re-opens 

1. Create a benchmark to run and verify IMDB. We weren't doing this before, the results were gathered using a script I set up here (https://github.com/Tmonster/verify_db_scripts). I plan on adding more to this repository soon. In order to properly verify with postgres, I had to modify the IMDB queries to return all lines (and not a min), with an order so that the COLLATE and the languages wouldn't produce differrent results.

2. Allow benchmarks to create and connect to a cached duckdb instance so that cached data doesn't need to be exported to parquet and imported for every query for every benchmark run. Results on improvements can be found in the table below. Timing information excludes the cost to download benchmark data, as that is very network dependent. 

A benchmark template can call connect {db_path}. The benchmark runner will then attempt to connect to the db at {db_path}. If an error is thrown, this is most likely because of storage format instability. When this is the case, the db is removed, and a connection is made to a clean instance. If no exception is thrown, the database is checked for tables. If no tables exist, data will be loaded in from duckdb_benchmark_data/{{cache}}.

If tables do exist, the benchmark will run as normal. If the data is incorrect, the benchmarks will report incorrect results fail. At this point it is up to the user to fix the data or answers, most likely by deleting the cached database.

Why do you need the full path from from the duckdb directory?
- If someone wants to connect to a duckdb that isn't in the duckdb_benchmark_data/{{cache}} folder, they can easily do that. 

| Benchmark  | Cached DB size | largest cached parquet file (uncompressed to CSV) | with "connect" | without "connect" |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| TPC-H (sf=1) | 262MB  | 719MB  | 9.507s  | 20.465s  |
| TPC-DS (sf=1)| 313MB  | 381MB  | 1:06 min  | 2:00 min  |
| clickbench  | 2.8GB  | 7.3G  | q19-q22 hang  | OOM locally  |
| h2oai-join  | 1.1G  | 440MB  | 19s  | 20.335s  |
| h2oai-group  | 584MB  | 466MB  | 17.139  | 18.806  |
| JOB  | 2GB | 915MB  | 3:11 min  | 7:05 min  |

TPC-H, TPC-DS, and IMDB all have improvements while h2oai regresses. I have noticed some weird behavior for clickbench where the size of the database grows between q19-q22 even though the queries are only select queries. Currently it hangs on my work computer, but I'll investigate next week on a more powerful computer.

 I could amend the code to open the connection in read only mode to prevent changes, but then it requires more logic when the db turns out to not have data and the benchmark runner needs to then delete the database and re-connect to load data.

### CODE REVIEW

**Comment by Tmonster** [comment]
IMDB regression fails because the old benchmark runner doesn't set a path for temporary storage, and the worker doesn't have enough memory to store the whole imdb database

**Comment by Tmonster** [comment]
IMDB regression fails because the old duckdb benchmark build can't load the imdb database into memory. When this merges, imdb regression tests should pass

File: benchmark/include/interpreted_benchmark.hpp [hunk 1]
```diff
@@ -15,6 +15,8 @@ namespace duckdb {
 struct BenchmarkFileReader;
 class MaterializedQueryResult;
 
+static const string DEFAULT_DB_PATH = "duckdb_benchmark_db.db";
```
Comment (Mytherin) [inline benchmark/include/interpreted_benchmark.hpp:18]: Is this used anywhere?

File: benchmark/interpreted_benchmark.cpp [hunk 2]
```diff
@@ -395,12 +412,13 @@ void InterpretedBenchmark::Cleanup(BenchmarkState *state_p) {
 }
 
 string InterpretedBenchmark::GetDatabasePath() {
-	if (!InMemory()) {
+	if (!InMemory() && cache_db.compare("") == 0) {
```
Comment (Mytherin) [inline benchmark/interpreted_benchmark.cpp:415]: nit: `cache_db.empty()`

File: benchmark/interpreted_benchmark.cpp [hunk 3]
```diff
@@ -395,12 +412,13 @@ void InterpretedBenchmark::Cleanup(BenchmarkState *state_p) {
...
 		string path = "duckdb_benchmark_db.db";
 		DeleteDatabase(path);
 		return path;
-	} else if (db_path != "") {
-		return db_path;
+	} else if (cache_db.compare("") != 0) {
```
Comment (Mytherin) [inline benchmark/interpreted_benchmark.cpp:419]: Nit: `cache_db.empty()`

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great. Some minor comments below.

> I have noticed some weird behavior for clickbench where the size of the database grows between q19-q22 even though the queries are only select queries.

Is that still happening? It might have something to do with the temporary table creation. I noticed earlier that creating temporary tables can trigger checkpoints in persistent databases (as the transaction mistakenly thinks changes have been made to the database).

**Comment by Tmonster** [comment]
@Mytherin From what I could tell the DB expansion doesn't happen anymore. The temp table issue fixed it

**Pull request closed without merging**